### PR TITLE
[#1010] Add Solr instructions for prebuilt Lucene index

### PR DESCRIPTION
**Pull request by adamyy**
https://github.com/castorini/anserini/issues/1010

* Provides instructions about how to configure Solr to read prebuilt Lucene indexes

### CODE REVIEW

File: docs/solrini.md [hunk 1]
```diff
@@ -98,6 +99,54 @@ Make sure `core18` collection is created and `/path/to/WashingtonPost` is update
 Solrini has also been verified to work with the [MS MARCO Passage Retrieval Corpus](experiments-msmarco-passage.md).
 There should be no major issues with other collections that are supported by Anserini, but we have not tested them.
 
+## Solr with Prebuilt Lucene Index
+
+Solr can be considered a front-end for Lucene, and it is entirely possible for Solr to read prebuilt Lucene indexes. 
+To achieve this, some housekeeping are required. The following uses [robust04](regressions-robust04.md) as an example.
```
Comment (lintool) [inline docs/solrini.md:105]: We adopt the convention of each sentence on a line; here and below.
robust04 -> Roboust04

File: docs/solrini.md [hunk 2]
```diff
@@ -98,6 +99,54 @@ Make sure `core18` collection is created and `/path/to/WashingtonPost` is update
...
+
+Second, make proper Solr schema adjustments if required.
+Here `robust04` is a Trec collection whose schema is already taken care of by [managed-schema](https://github.com/castorini/anserini/blob/master/src/main/resources/solr/anserini/conf/managed-schema) in the Solr configset.
+However, if you are dealing with a collection such as `cord19`, remember to make proper adjustments to the Solr schema, as [previously described](#setting-up-a-single-node-solrcloud-instance).
+
+```
+curl -X POST -H 'Content-type:application/json' --data-binary @src/main/resources/solr/schemas/SCHEMA_NAME.json http://localhost:8983/solr/COLLECTION_NAME/schema
+```
+
+Then, either *copy the index files to where Solr expected*, or *tell Solr where to look for*.
```
Comment (lintool) [inline docs/solrini.md:124]: Let's be opinionated - pick the better option (my vote is for the first) and do that - don't give the user the choice.

File: docs/solrini.md [hunk 3]
```diff
@@ -98,6 +99,54 @@ Make sure `core18` collection is created and `/path/to/WashingtonPost` is update
...
+
+```
+solrini/bin/solr create -n anserini -c robust04
+```
+
+Along with the collection, Solr will create a core instance, whose name can be found in the Solr UI under collection overview.
+Which might look something like `<collection_name>_shard<id>_replica_<id>` (e.g., `robust04_shard1_replica_n1`). Solr stores configurations and data for the core instances under Solr home, which for us is `solrini/server/solr/` by default.
+
+Second, make proper Solr schema adjustments if required.
+Here `robust04` is a Trec collection whose schema is already taken care of by [managed-schema](https://github.com/castorini/anserini/blob/master/src/main/resources/solr/anserini/conf/managed-schema) in the Solr configset.
```
Comment (lintool) [inline docs/solrini.md:117]: Trec -> TREC

**Review by lintool — CHANGES_REQUESTED** [review]
Minor comments, thanks for the PR!

**Comment by adamyy** [comment]
Actually, as I was testing the prebuilt index with `run_solr_integration.py`, I noticed that there are discrepancy between expected and actual trec_eval benchmark. Currently looking into the cause and this PR might not be ready to merge just yet!

**Comment by adamyy** [comment]
Looked a bit more into this, turns out my index was not built with the `contents` field (only `id` and `raw`) and `SearchSolr` uses `contents` as the default field of search. So it either errors out / produce nonsense output.

I guess it is worth noting that the prebuilt index should have the `contents` field for this to work properly? When we are indexing into Solr using `IndexCollection`, it seems like the `contents` field are created regardless of whether we specify `-storeContents` or not. However, if we are just producing a Lucene index then `contents` field only present if we specify `-storeContents`.

**Comment by lintool** [comment]
yup, this should be documented somewhere!

**Comment by lintool** [comment]
Also, check out: https://github.com/castorini/anserini/blob/master/.travis.yml#L20

It _might_ be reasonable to do a complete e2e integration test based on this? I mean, travis is running it all the time... just a matter of moving it inside JUnit?

**Comment by adamyy** [comment]
Actually, I have misunderstood the issue again. 😅 

I read more about what `stored` means for a Lucene Document Field, and found this [explanation](https://stackoverflow.com/a/4185226). Turns out `stored=false` still makes the field searchable, it is just that the field is not stored as is, and will not be returned as part of the query result.

Then I was wondering, how come `SearchCollection` on non-`storeContents` MSMARCO index can produce the output that matches the regression, but `SearchSolr` cannot?

I noticed [solrconfig.xml#L785](https://github.com/castorini/anserini/blob/master/src/main/resources/solr/anserini/conf/solrconfig.xml#L785) set the default field of search to a field called `_text_` (this is the default Solr config). However, in our case we want it to be `contents`, and it is what we were trying to enforce at [SearchSolr.java#L225](https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/search/SearchSolr.java#L225). My hunch is that Solr internally picks the default field of search based on both query parameters and solrconfig.xml, and it might have picked the wrong one in my previous run.

I modified this field from `_text_` to `contents`, recreated the Solr core with the updated configset and the non-`storeContents` index, then performed `SearchSolr` again. The good news is that the result now matches the regression. I will perform a couple more tests to ensure that this is what is causing the issue.

**Comment by adamyy** [comment]
> Also, check out: https://github.com/castorini/anserini/blob/master/.travis.yml#L20
> 
> It _might_ be reasonable to do a complete e2e integration test based on this? I mean, travis is running it all the time... just a matter of moving it inside JUnit?

Do we want to do `run_solr_regression.py` in travis? @lintool

**Comment by adamyy** [comment]
> I modified this field from `_text_` to `contents`, recreated the Solr core with the updated configset and the non-`storeContents` index, then performed `SearchSolr` again. The good news is that the result now matches the regression. I will perform a couple more tests to ensure that this is what is causing the issue.

Big oops, I tested my assumption again with a newly built index (without `-storeContents` flag) everything worked just fine. TL;DR I think my original index was broken. Sorry for causing the confusion! This PR should be ready now! @lintool

**Comment by lintool** [comment]
> Do we want to do `run_solr_regression.py` in travis? @lintool

Maybe `run_solr_regression.py` with the small CACM collection in travis? Shouldn't be too difficult to do? If you think it's a good idea, create another issue to keep track of progress?

**Review by lintool — APPROVED** [review]

**Pull request merged**